File: src/main/java/io/github/zenwave360/zdl/antlr/ZdlModel.java
package io.github.zenwave360.zdl.antlr;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;

public class ZdlModel extends FluentMap {
    public ZdlModel() {
        put("imports", new ArrayList<>());
        put("config", new FluentMap());
        put("apis", new FluentMap());
        put("aggregates", new FluentMap());
        put("entities", new FluentMap());
        put("enums", new FluentMap());
        put("relationships", new FluentMap());
        put("services", new FluentMap());
        put("inputs", new FluentMap());
        put("outputs", new FluentMap());
        put("events", new FluentMap());
        put("locations", new FluentMap());
        put("problems", new ArrayList<>());
    }

    public FluentMap getAggregates() {
        return (FluentMap) get("aggregates");
    }

    public FluentMap getEntities() {
        return (FluentMap) get("entities");
    }

    public FluentMap getInputs() {
        return (FluentMap) get("inputs");
    }

    public FluentMap getOutputs() {
        return (FluentMap) get("outputs");
    }

    public FluentMap getEvents() {
        return (FluentMap) get("events");
    }


    public FluentMap getEnums() {
        return (FluentMap) get("enums");
    }

    public FluentMap getRelationships() {
        return (FluentMap) get("relationships");
    }

    public FluentMap getLocations() {
        return (FluentMap) get("locations");
    }

    public List getProblems() {
        return (List) get("problems");
    }

    public FluentMap setLocation(String location, int[] locations) {
        if(locations == null || locations.length != 6) {
            return this;
        }
        return appendTo("locations", location, locations);
    }

    public void clearProblems() {
        if(getProblems() != null) {
            getProblems().clear();
        }
    }
    public void addProblem(String path, String value, String error) {
        try {
            appendToList("problems", problem(path, value, error));
        } catch (Exception e) {
            System.err.printf("Error adding problem '%s': %s%n", path, e.getMessage());
        }
    }

    private Map problem(String path, String value, String error) {
        int[] location = getLocation(path);
        return Map.of(
                "path", path,
                "location", location,
                "value", value,
                "message", String.format(error, value)
        );
    }

    private int[] getLocation(String path) {
        return JSONPath.get(this, "$.locations.['" + path + "']");
    }

    public String getLocation(int line, int character) {
        var locations = getLocations().entrySet().stream().filter(k -> {
            var position = (int[]) k.getValue();
            var lineStart = position[2];
            var characterStart = position[3];
            var lineEnd = position[4];
            var characterEnd = position[5];
            return lineStart <= line && line <= lineEnd && (line != lineStart || characterStart <= character) && (line != lineEnd || character <= characterEnd);
        }).toList();
        // return the one with the closer location which is the one with the smaller range
        var location = locations.stream().min(Comparator.comparingInt(k -> {
            var position = (int[]) k.getValue();
            return position[1] - position[0];
        })).map(Map.Entry::getKey).orElse(null);
        return location;
    }
}

File: src/commonMain/kotlin/io/zenwave360/zdl/antlr/ZdlModel.kt
package io.zenwave360.zdl.antlr

class ZdlModel(private val delegate: FluentMap = FluentMap.build()) : MutableMap<String, Any?> by delegate {

    init {
        // Initialize top-level structure
        delegate.putEntry("imports", mutableListOf<Any?>())
        delegate.putEntry("config", FluentMap.build())
        delegate.putEntry("apis", FluentMap.build())
        delegate.putEntry("aggregates", FluentMap.build())
        delegate.putEntry("entities", FluentMap.build())
        delegate.putEntry("enums", FluentMap.build())
        delegate.putEntry("relationships", FluentMap.build())
        delegate.putEntry("services", FluentMap.build())
        delegate.putEntry("inputs", FluentMap.build())
        delegate.putEntry("outputs", FluentMap.build())
        delegate.putEntry("events", FluentMap.build())
        delegate.putEntry("locations", FluentMap.build())
        delegate.putEntry("problems", mutableListOf<Any?>())
    }

    // Forward FluentMap fluent API expected by call sites
    fun with(key: String, value: Any?): ZdlModel { delegate.with(key, value); return this }
    fun appendTo(collection: String, key: String, value: Any?): ZdlModel { delegate.appendTo(collection, key, value); return this }
    fun appendTo(collection: String, map: Map<String, Any?>): ZdlModel { delegate.appendTo(collection, map); return this }
    fun appendToList(collection: String, value: Any?): ZdlModel { delegate.appendToList(collection, value); return this }

    fun getAggregates(): FluentMap = delegate["aggregates"] as FluentMap
    fun getEntities(): FluentMap = delegate["entities"] as FluentMap
    fun getInputs(): FluentMap = delegate["inputs"] as FluentMap
    fun getOutputs(): FluentMap = delegate["outputs"] as FluentMap
    fun getEvents(): FluentMap = delegate["events"] as FluentMap
    fun getEnums(): FluentMap = delegate["enums"] as FluentMap
    fun getRelationships(): FluentMap = delegate["relationships"] as FluentMap
    fun getLocations(): FluentMap = delegate["locations"] as FluentMap

    @Suppress("UNCHECKED_CAST")
    fun getProblems(): MutableList<MutableMap<String, Any?>> =
        delegate["problems"] as MutableList<MutableMap<String, Any?>>

    fun setLocation(location: String, positions: IntArray?): FluentMap {
        if (positions == null || positions.size != 6) return getLocations()
        val locations = getLocations()
        locations.appendTo(location, location, positions)
        return locations
    }

    fun clearProblems() = getProblems().clear()

    fun addProblem(path: String, value: String?, error: String) {
        val p = problem(path, value, error)
        @Suppress("UNCHECKED_CAST")
        (delegate["problems"] as MutableList<Any?>).add(p)
    }

    private fun problem(path: String, value: String?, error: String): Map<String, Any?> {
        val location = getLocation(path)
        val message = error.replace("%s", value ?: "")
        return mapOf(
            "path" to path,
            "location" to location,
            "value" to value,
            "message" to message
        )
    }

    private fun getLocation(path: String): IntArray? {
        @Suppress("UNCHECKED_CAST")
        return JSONPath.get(this, "$.locations.['$path']") as? IntArray
    }

    fun getLocation(line: Int, character: Int): String? {
        val entries = getLocations().entries.filter { (_, value) ->
            val position = value as IntArray
            val lineStart = position[2]
            val characterStart = position[3]
            val lineEnd = position[4]
            val characterEnd = position[5]
            lineStart <= line && line <= lineEnd &&
                (line != lineStart || characterStart <= character) &&
                (line != lineEnd || character <= characterEnd)
        }
        val location = entries.minByOrNull { (_, v) ->
            val pos = v as IntArray
            pos[1] - pos[0]
        }?.key
        return location
    }
}

